typescript/infra/src/utils/helm.ts

```diff
@@ -91,3 +92,97 @@ export function getDeployableHelmChartName(helmChartConfig: HelmChartConfig) {
 export function buildHelmChartDependencies(chartPath: string) {
   return execCmd(`cd ${chartPath} && helm dependency build`, {}, false, true);
 }
+
+export type HelmValues = Record<string, any>;
+
+export abstract class HelmManager<T = HelmValues> {
+  abstract readonly helmReleaseName: string;
+  abstract readonly helmChartPath: string;
+  abstract readonly namespace: string;
+
+  /**
+   * Returns the values to be passed to the helm chart.
+   * Expected to be an object of values.
+   */
+  abstract helmValues(): Promise<T>;
+
+  async runHelmCommand(action: HelmCommand, dryRun?: boolean): Promise<void> {
+    const cmd = ['helm', action];
+    if (dryRun) cmd.push('--dry-run');
+
+    if (action == HelmCommand.Remove) {
+      if (dryRun) cmd.push('--dry-run');
+      cmd.push(this.helmReleaseName, this.namespace);
+      await execCmd(cmd, {}, false, true);
+      return;
+    }
+
+    const values = helmifyValues(await this.helmValues());
+    if (action == HelmCommand.InstallOrUpgrade && !dryRun) {
+      // Delete secrets to avoid them being stale
+      const cmd = [
+        'kubectl',
+        'delete',
+        'secrets',
+        '--namespace',
+        this.namespace,
+        '--selector',
+        `app.kubernetes.io/instance=${this.helmReleaseName}`,
+      ];
+      try {
+        await execCmd(cmd, {}, false, false);
+      } catch (e) {
+        console.error(e);
+      }
+    }
+
+    await buildHelmChartDependencies(this.helmChartPath);
+    cmd.push(
+      this.helmReleaseName,
+      this.helmChartPath,
+      '--create-namespace',
+      '--namespace',
+      this.namespace,
+      ...values,
+    );
+    if (action == HelmCommand.UpgradeDiff) {
+      cmd.push(
+        `| kubectl diff --namespace ${this.namespace} --field-manager="Go-http-client" -f - || true`,
+      );
+    }
+    await execCmd(cmd, {}, false, true);
+  }
+
+  async doesHelmReleaseExist() {
+    try {
+      await execCmd(
+        `helm status ${this.helmReleaseName} --namespace ${this.namespace}`,
+        {},
+        false,
+        false,
+      );
+      return true;
+    } catch (error) {
+      return false;
+    }
+  }
+
+  async getExistingK8sSecrets(): Promise<string[]> {
+    const [output] = await execCmd(
+      `kubectl get secret --selector=app.kubernetes.io/instance=${this.helmReleaseName} -o jsonpath='{.items[*].metadata.name}' -n ${this.namespace}`,
```

Comment:
Could look at using https://www.npmjs.com/package/@kubernetes/client-node in the future

Reply:
yeah would support this for sure